cranelift/codegen/src/isle_prelude.rs

```diff
@@ -81,27 +81,43 @@ macro_rules! isle_common_prelude_methods {
         }
 
         #[inline]
-        fn u64_shl(&mut self, x: u64, y: u64) -> u64 {
-            x << y
+        fn imm64_shl(&mut self, ty: Type, x: Imm64, y: Imm64) -> Imm64 {
+            // Mask off any excess shift bits.
+            let shift_mask = (ty.bits() - 1) as u64;
+            let y = (y.bits() as u64) & shift_mask;
+
+            // Mask the result to `ty` bits.
+            let ty_mask = u64::MAX >> u32::checked_sub(64, ty.bits()).unwrap_or(0);
```

Comment:
We have a `ty_mask` function in the ISLE prelude that I think we could delegate to here instead?